File: ios/Classes/Handlers/VideoPlayerObserver.swift
import AVFoundation
import Foundation

extension VideoPlayerView {
    /// Adds item-level KVO observers and notification observers to a player item.
    /// Use this when replacing an item mid-session (e.g., stall recovery) where
    /// player-level observers are already registered.
    func addItemObservers(to item: AVPlayerItem) {
        // Clear per-item error latches so each new item starts clean. Can't
        // rely solely on the .readyToPlay branch — an item that fails during
        // manifest load transitions .unknown → .failed directly and would
        // otherwise inherit the previous item's suppression flag.
        errorEmittedForCurrentItem = false
        lastErrorStatusCode = 0

        item.addObserver(self, forKeyPath: "status", options: [.new, .old], context: nil)
        item.addObserver(self, forKeyPath: "playbackBufferEmpty", options: [.new], context: nil)
        item.addObserver(self, forKeyPath: "playbackLikelyToKeepUp", options: [.new], context: nil)
        item.addObserver(self, forKeyPath: "playbackBufferFull", options: [.new], context: nil)

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemFailedToPlay),
            name: .AVPlayerItemFailedToPlayToEndTime,
            object: item
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemDidStall),
            name: AVPlayerItem.playbackStalledNotification,
            object: item
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(playerItemNewErrorLogEntry),
            name: .AVPlayerItemNewErrorLogEntry,
            object: item
        )

        NotificationCenter.default.addObserver(
            self,
            selector: #selector(videoDidEnd),
            name: .AVPlayerItemDidPlayToEndTime,
            object: item
        )
    }

    /// Adds all observers (item-level + player-level) for initial setup.
    /// Player-level observers are only added once per view lifetime via hasPlayerObservers guard.
    func addObservers(to item: AVPlayerItem) {
        addItemObservers(to: item)

        // Player-level observers should only be added once. Calling addObservers
        // multiple times (e.g., on re-load) would otherwise register duplicates,
        // causing double event emission.
        if !hasPlayerObservers {
            player?.addObserver(self, forKeyPath: "timeControlStatus", options: [.new, .old], context: nil)
            player?.addObserver(self, forKeyPath: "externalPlaybackActive", options: [.new, .initial], context: nil)
            hasPlayerObservers = true
        }

        // Audio route change is a global notification — only need one observer per view
        NotificationCenter.default.removeObserver(self, name: AVAudioSession.routeChangeNotification, object: nil)
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(handleAudioRouteChange),
            name: AVAudioSession.routeChangeNotification,
            object: nil
        )
    }

    public override func observeValue(
        forKeyPath keyPath: String?,
        of object: Any?,
        change: [NSKeyValueChangeKey: Any]?,
        context: UnsafeMutableRawPointer?
    ) {
        // Handle AVPlayerItem observations
        if let item = object as? AVPlayerItem {
            // Ignore callbacks from a replaced/stale item to prevent phantom state changes
            guard item === player?.currentItem else { return }
            switch keyPath {
            case "status":
                switch item.status {
                case .readyToPlay:
                    // Only send isInitialized for new players, not for shared players
                    // Shared players already sent their state in the init
                    if !isSharedPlayer {
                        sendEvent("isInitialized")
                    }
                case .failed:
                    emitErrorOnce(message: item.error?.localizedDescription ?? "Unknown")
                default: break
                }
            case "playbackBufferEmpty":
                // Only send buffering event when buffer is empty AND playback has stalled
                // This prevents false buffering events when the player has enough buffer to continue
                if item.isPlaybackBufferEmpty, let player = player {
                    // Only send buffering if the player is waiting to play (actually stalled)
                    // or if we're seeking (reasonForWaitingToPlay is not nil)
                    if player.timeControlStatus == .waitingToPlayAtSpecifiedRate {
                        sendEvent("buffering")
                    }
                }
            case "playbackLikelyToKeepUp":
                if item.isPlaybackLikelyToKeepUp {
                    if let player = player {
                        if player.timeControlStatus == .waitingToPlayAtSpecifiedRate {
                            // Buffer is ready but player is stuck waiting — seek to
                            // live edge to force video decoder keyframe re-sync, then play.
                            seekToLiveEdgeAndPlay()
                        } else if player.rate > 0 && player.timeControlStatus == .playing {
                            sendEvent("play")
                        } else if player.timeControlStatus == .paused && player.reasonForWaitingToPlay == nil {
                            if !isPipCurrentlyActive {
                                sendEvent("pause")
                            }
                        }
                    }
                }
            case "playbackBufferFull":
                // Secondary recovery: playbackLikelyToKeepUp can stay false even when
                // buffer is full (confirmed Apple docs). Force playback if stuck.
                if item.isPlaybackBufferFull, let player = player,
                   player.timeControlStatus == .waitingToPlayAtSpecifiedRate {
                    seekToLiveEdgeAndPlay()
                }
            default: break
            }
        } else if let observedPlayer = object as? AVPlayer, observedPlayer == player {
            // Handle AVPlayer observations
            switch keyPath {
            case "timeControlStatus":
                guard let player = player else { return }

                switch player.timeControlStatus {
                case .playing:
                    // ALWAYS update Now Playing info when playback starts
                    // This ensures media controls show the correct info whether in normal view or PiP
                    var mediaInfo = currentMediaInfo

                    // Fallback: Try to retrieve from SharedPlayerManager if not available locally
                    if mediaInfo == nil, let controllerIdValue = controllerId {
                        mediaInfo = SharedPlayerManager.shared.getMediaInfo(for: controllerIdValue)
                        if mediaInfo != nil {
                            currentMediaInfo = mediaInfo // Update local copy
                        }
                    }

                    if let mediaInfo = mediaInfo {
                        setupNowPlayingInfo(mediaInfo: mediaInfo)
                    }

                    // Enable automatic PiP when playback starts (even from native controls)
                    // This ensures auto PiP works whether the user taps Flutter controls or native controls
                    if #available(iOS 14.2, *) {
                        if let controllerIdValue = controllerId {
                            // Check if there's already a primary view for this controller
                            let hasPrimaryView = SharedPlayerManager.shared.getPrimaryViewId(for: controllerIdValue) != nil

                            if !hasPrimaryView {
                                // No primary view set yet - this means the user started playback via native controls
                                // Set THIS view as primary
                                SharedPlayerManager.shared.setPrimaryView(viewId, for: controllerIdValue)
                            }

                            // Check if THIS view is the primary view for this controller
                            if SharedPlayerManager.shared.isPrimaryView(viewId, for: controllerIdValue) {
                                // For shared players, check the shared settings instead of instance variable
                                // This ensures the second view uses the same PiP settings as the first view
                                let shouldEnableAutoPiP: Bool
                                if let sharedSettings = SharedPlayerManager.shared.getPipSettings(for: controllerIdValue) {
                                    shouldEnableAutoPiP = sharedSettings.canStartPictureInPictureAutomatically
                                } else {
                                    shouldEnableAutoPiP = canStartPictureInPictureAutomatically
                                }

                                if shouldEnableAutoPiP {
                                    SharedPlayerManager.shared.setAutomaticPiPEnabled(for: controllerIdValue, enabled: true)

                                    // Eagerly create the PiP controller so it's warmed up
                                    // for background PiP (freshly created controllers may
                                    // not be ready for immediate startPictureInPicture).
                                    if #available(iOS 14.0, *) {
                                        ensurePipController()
                                    }

                                    // Ensure media info is set again after enabling PiP
                                    // This guarantees media controls work correctly in PiP mode
                                    if let mediaInfo = currentMediaInfo {
                                        setupNowPlayingInfo(mediaInfo: mediaInfo)
                                    }
                                }
                            }
                        }
                    }

                    sendEvent("play")
                case .paused:
                    if player.reasonForWaitingToPlay == nil && !isPipCurrentlyActive {
                        // With automaticallyWaitsToMinimizeStalling = false,
                        // a silent buffer exhaustion drops rate to 0 here
                        // rather than going through .waitingToPlayAtSpecifiedRate.
                        // Inspect the item's buffer state so stall recovery
                        // engages — but skip the check when the user just
                        // hit pause, since a tight live buffer can read as
                        // "not likely to keep up" even on a deliberate pause.
                        let item = player.currentItem
                        let stalled = !userRequestedPause
                            && (item?.isPlaybackBufferEmpty == true
                                || item?.isPlaybackLikelyToKeepUp == false)
                        if stalled, let item = item, item.status != .failed {
                            sendEvent("buffering")
                        } else {
                            sendEvent("pause")
                        }
                    }
                case .waitingToPlayAtSpecifiedRate:
                    // With automaticallyWaitsToMinimizeStalling = false (set for live),
                    // the reason is .evaluatingBufferingRate, never .toMinimizeStalls.
                    // Send buffering for any waiting state to ensure Dart knows we're stalled.
                    sendEvent("buffering")
                @unknown default:
                    break
                }
            case "externalPlaybackActive":
                guard let player = player else { return }
                let isActive = player.isExternalPlaybackActive

                if isActive {
                    // When AirPlay connects, try to get device name with multiple retry attempts

                    // Try to get device name immediately
                    let deviceName = getAirPlayDeviceName()

                    // Send initial event (might have deviceName or might be nil)
                    var eventData: [String: Any] = ["isConnected": isActive, "isConnecting": false]
                    if let deviceName = deviceName {
                        eventData["deviceName"] = deviceName
                    }

                    // Send through per-view event channel (legacy)
                    sendEvent("airPlayConnectionChanged", data: eventData)

                    // Send through controller-level event channel (persists when views disposed)
                    if let controllerIdValue = controllerId {
                        SharedPlayerManager.shared.sendControllerEvent(
                            "airPlayConnectionChanged",
                            data: eventData,
                            for: controllerIdValue
                        )
                    }

                    // If device name is nil, retry multiple times with increasing delays
                    if deviceName == nil {
                        retryGetAirPlayDeviceName(attempt: 1, maxAttempts: 4)
                    }
                } else {
                    // Disconnected from AirPlay
                    var eventData: [String: Any] = ["isConnected": false, "isConnecting": false]

                    // Send through per-view event channel (legacy)
                    sendEvent("airPlayConnectionChanged", data: eventData)

                    // Send through controller-level event channel (persists when views disposed)
                    if let controllerIdValue = controllerId {
                        SharedPlayerManager.shared.sendControllerEvent(
                            "airPlayConnectionChanged",
                            data: eventData,
                            for: controllerIdValue
                        )
                    }
                }
            default: break
            }
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }

    @objc func playerItemDidStall(notification: Notification) {
        sendEvent("buffering")
        // Attempt recovery after brief delay — seek to live edge to force
        // video decoder keyframe re-sync (plain play() can leave video frozen).
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            guard let self = self, let player = self.player else { return }
            if player.timeControlStatus != .playing {
                self.seekToLiveEdgeAndPlay()
            }
        }
    }

    /// Seeks to the live edge and starts playback. Forces both audio and video
    /// decoders to re-sync to a fresh keyframe, preventing frozen-video-with-audio.
    /// Guarded against re-entrancy — multiple recovery paths can fire concurrently
    /// and overlapping seeks would cause thrashing.
    func seekToLiveEdgeAndPlay() {
        guard !isRecoveringFromStall else { return }
        isRecoveringFromStall = true

        guard let player = player, let item = player.currentItem else {
            player?.play()
            isRecoveringFromStall = false
            return
        }

        if let lastRange = item.seekableTimeRanges.last?.timeRangeValue {
            // Seek to ~2s before the live edge to leave buffer headroom.
            // Seeking to the exact end leaves zero buffer ahead, causing
            // immediate re-buffering on any network fluctuation.
            let liveEdge = CMTimeRangeGetEnd(lastRange)
            let buffer = CMTimeMake(value: 2, timescale: 1)
            let target = CMTimeSubtract(liveEdge, buffer)
            let seekTarget = CMTimeMaximum(target, lastRange.start)
            player.seek(to: seekTarget, toleranceBefore: .zero, toleranceAfter: .zero) { [weak self] _ in
                self?.player?.play()
                self?.isRecoveringFromStall = false
            }
        } else if isPipCurrentlyActive {
            // During PiP, don't replace the AVPlayerItem — item replacement
            // can cause PiP to close. Just try to resume playback and let
            // the existing item recover.
            player.play()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.isRecoveringFromStall = false
            }
        } else {
            // seekableTimeRanges is empty — the player is too far behind the live
            // window or the manifest hasn't loaded yet. Replace the AVPlayerItem
            // to force a full re-fetch of the manifest and segments.
            if let asset = item.asset as? AVURLAsset {
                let freshItem = AVPlayerItem(asset: AVURLAsset(url: asset.url))
                self.removeItemObservers(from: item)
                player.replaceCurrentItem(with: freshItem)
                // Only add item-level observers — player-level are already registered
                self.addItemObservers(to: freshItem)
                player.play()
            } else {
                player.play()
            }
            // Delay clearing the re-entrancy guard so KVO callbacks from the
            // fresh item don't immediately re-trigger recovery while it loads.
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.isRecoveringFromStall = false
            }
        }
    }

    /// Removes KVO observers and notification observers from a player item.
    /// Must mirror exactly what `addItemObservers(to:)` registers.
    func removeItemObservers(from item: AVPlayerItem) {
        item.removeObserver(self, forKeyPath: "status")
        item.removeObserver(self, forKeyPath: "playbackBufferEmpty")
        item.removeObserver(self, forKeyPath: "playbackLikelyToKeepUp")
        item.removeObserver(self, forKeyPath: "playbackBufferFull")
        NotificationCenter.default.removeObserver(self, name: .AVPlayerItemFailedToPlayToEndTime, object: item)
        NotificationCenter.default.removeObserver(self, name: AVPlayerItem.playbackStalledNotification, object: item)
        NotificationCenter.default.removeObserver(self, name: .AVPlayerItemNewErrorLogEntry, object: item)
        NotificationCenter.default.removeObserver(self, name: .AVPlayerItemDidPlayToEndTime, object: item)
    }

    @objc func playerItemFailedToPlay(notification: Notification) {
        let message = (notification.userInfo?[AVPlayerItemFailedToPlayToEndTimeErrorKey] as? Error)?
            .localizedDescription ?? "Unknown error"
        emitErrorOnce(message: message)
    }

    @objc func playerItemNewErrorLogEntry(notification: Notification) {
        guard let item = notification.object as? AVPlayerItem,
              let errorLog = item.errorLog(),
              let lastEvent = errorLog.events.last else { return }
        lastErrorStatusCode = lastEvent.errorStatusCode
        NSLog("[VideoPlayer] Error log: status=\(lastEvent.errorStatusCode) domain=\(lastEvent.errorDomain) comment=\(lastEvent.errorComment ?? "none") URI=\(lastEvent.uri ?? "none")")
    }

    /// Emits an "error" event at most once per AVPlayerItem. AVFoundation
    /// fires both `item.status = .failed` and `AVPlayerItemFailedToPlayToEndTime`
    /// for the same underlying failure; without this guard Dart runs the
    /// recovery ladder twice and burns a refresh-attempt slot.
    func emitErrorOnce(message: String) {
        if errorEmittedForCurrentItem { return }
        errorEmittedForCurrentItem = true
        var data: [String: Any] = ["message": message]
        if lastErrorStatusCode != 0 { data["statusCode"] = lastErrorStatusCode }
        sendEvent("error", data: data)
    }

    @objc func videoDidEnd() {
        if enableLooping {
            // For smooth looping, seek to beginning and continue playing
            player?.seek(to: .zero) { [weak self] finished in
                if finished {
                    // Continue playing for seamless loop
                    self?.player?.play()
                }
            }
            // Don't send completed event when looping to match Android behavior
            // (Android with REPEAT_MODE_ONE doesn't reach STATE_ENDED)
        } else {
            // Reset video to the beginning and pause
            player?.seek(to: .zero)
            player?.pause()
            sendEvent("completed")
        }
    }

    // MARK: - AirPlay Route Detection
    //
    // Per-view AVRouteDetector was removed — all route detection is owned by
    // SharedPlayerManager.startAirPlayRouteDetection(). Apple warns the
    // detector "significantly increases power consumption", so a single
    // shared instance is preferred over one per view.

    /// Gets the name of the currently connected AirPlay device
    func getAirPlayDeviceName() -> String? {
        let currentRoute = AVAudioSession.sharedInstance().currentRoute
        for output in currentRoute.outputs {
            if output.portType == .airPlay {
                return output.portName
            }
        }
        return nil
    }

    /// Retries getting the AirPlay device name with exponential backoff
    ///
    /// This function recursively retries getting the device name because iOS sometimes
    /// takes time to update the audio route when AirPlay video streaming starts.
    ///
    /// - Parameters:
    ///   - attempt: Current attempt number (1-based)
    ///   - maxAttempts: Maximum number of retry attempts
    func retryGetAirPlayDeviceName(attempt: Int, maxAttempts: Int) {
        guard attempt <= maxAttempts else {
            return
        }

        // Calculate delay with exponential backoff: 0.1s, 0.3s, 0.6s, 1.0s
        let delay: Double
        switch attempt {
        case 1: delay = 0.1
        case 2: delay = 0.3
        case 3: delay = 0.6
        default: delay = 1.0
        }


        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self = self else { return }

            let deviceName = self.getAirPlayDeviceName()

            if let deviceName = deviceName {
                // Success! Send event with device name
                var eventData: [String: Any] = ["isConnected": true, "isConnecting": false]
                eventData["deviceName"] = deviceName

                // Send through per-view event channel (legacy)
                self.sendEvent("airPlayConnectionChanged", data: eventData)

                // Send through controller-level event channel (persists when views disposed)
                if let controllerIdValue = self.controllerId {
                    SharedPlayerManager.shared.sendControllerEvent(
                        "airPlayConnectionChanged",
                        data: eventData,
                        for: controllerIdValue
                    )
                }
            } else if attempt < maxAttempts {
                // Try again
                self.retryGetAirPlayDeviceName(attempt: attempt + 1, maxAttempts: maxAttempts)
            } else {
                // Exhausted all retries
                // Send event without device name - the Dart caching layer will handle it
                var eventData: [String: Any] = ["isConnected": true, "isConnecting": false]

                // Send through per-view event channel (legacy)
                self.sendEvent("airPlayConnectionChanged", data: eventData)

                // Send through controller-level event channel (persists when views disposed)
                if let controllerIdValue = self.controllerId {
                    SharedPlayerManager.shared.sendControllerEvent(
                        "airPlayConnectionChanged",
                        data: eventData,
                        for: controllerIdValue
                    )
                }
            }
        }
    }

    /// Handles audio route changes to detect AirPlay device changes
    @objc func handleAudioRouteChange(notification: Notification) {

        // Log the reason for the route change
        if let reason = notification.userInfo?[AVAudioSessionRouteChangeReasonKey] as? UInt {
            let reasonString: String
            switch AVAudioSession.RouteChangeReason(rawValue: reason) {
            case .newDeviceAvailable: reasonString = "NewDeviceAvailable"
            case .oldDeviceUnavailable: reasonString = "OldDeviceUnavailable"
            case .categoryChange: reasonString = "CategoryChange"
            case .override: reasonString = "Override"
            case .wakeFromSleep: reasonString = "WakeFromSleep"
            case .noSuitableRouteForCategory: reasonString = "NoSuitableRouteForCategory"
            case .routeConfigurationChange: reasonString = "RouteConfigurationChange"
            default: reasonString = "Unknown(\(reason))"
            }
        }

        guard let player = player else { return }

        let deviceName = getAirPlayDeviceName()
        let isPlayerActive = player.isExternalPlaybackActive

        // Use same logic as initial state check:
        // We're connected if EITHER the player is using AirPlay OR device is in audio route
        let isSystemActive = deviceName != nil
        let isConnected = isPlayerActive || isSystemActive

        // Determine if we're in a connecting state:
        // - AirPlay device is present in audio route (systemActive)
        // - But player hasn't started streaming yet (!playerActive)
        // - AND we consider this "connected" at system level (isConnected)
        let isConnecting = isSystemActive && !isPlayerActive && isConnected

        // Only send events for AirPlay-related changes
        if deviceName != nil || isPlayerActive {

            var eventData: [String: Any] = [
                "isConnected": isConnected,
                "isConnecting": isConnecting
            ]
            if let deviceName = deviceName {
                eventData["deviceName"] = deviceName
            }

            // Send through per-view event channel (legacy)
            sendEvent("airPlayConnectionChanged", data: eventData)

            // Send through controller-level event channel (persists when views disposed)
            if let controllerIdValue = controllerId {
                SharedPlayerManager.shared.sendControllerEvent(
                    "airPlayConnectionChanged",
                    data: eventData,
                    for: controllerIdValue
                )
            }
        } else {
        }
    }
}

File: android/src/main/kotlin/com/huddlecommunity/better_native_video_player/handlers/VideoPlayerObserver.kt
package com.huddlecommunity.better_native_video_player.handlers

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.Timeline
import androidx.media3.common.C

/**
 * Observes ExoPlayer state changes and reports them via EventHandler
 * Equivalent to iOS VideoPlayerObserver
 */
class VideoPlayerObserver(
    private val player: Player,
    private val eventHandler: VideoPlayerEventHandler,
    private val notificationHandler: com.huddlecommunity.better_native_video_player.handlers.VideoPlayerNotificationHandler? = null,
    private val getMediaInfo: (() -> Map<String, Any>?)? = null,
    private val controllerId: Int? = null,
    private val viewId: Long? = null,
    private val isInPipMode: () -> Boolean = { false }
) : Player.Listener {

    companion object {
        private const val TAG = "VideoPlayerObserver"
        private const val UPDATE_INTERVAL_MS = 1000L
        private const val STALL_WATCHDOG_MS = 12_000L
        private const val MAX_RECOVERY_ATTEMPTS = 3
    }

    // Track if we've already sent a buffering event to avoid duplicates
    private var hasReportedBuffering = false

    // Track Cast/external playback connection state
    private var wasExternalPlaybackActive = false

    // Reusable objects to avoid per-cycle allocations on the main thread
    private val handler = Handler(Looper.getMainLooper())
    private val timelineWindow = Timeline.Window()

    // Stall watchdog: detects indefinite STATE_BUFFERING and forces recovery
    private var stallWatchdogRunnable: Runnable? = null
    private var stallRecoveryAttempt = 0

    // Guard to start timeUpdateRunnable only once (on first STATE_READY)
    private var isTimeUpdateRunning = false

    private val timeUpdateRunnable = object : Runnable {
        override fun run() {
            var position: Long
            var duration: Long

            val timeline = player.currentTimeline

            // getWindow populates timelineWindow and returns it — single call
            val isLiveStream = !timeline.isEmpty &&
                timeline.getWindow(player.currentMediaItemIndex, timelineWindow).isDynamic &&
                !timelineWindow.isSeekable

            if (isLiveStream) {
                duration = timelineWindow.durationMs
                // currentPosition is already relative to the window start —
                // windowStartTimeMs is a Unix-epoch timestamp and must not
                // be subtracted from it.
                position = player.currentPosition

                if (position < 0) position = 0
                if (position > duration) position = duration
            } else {
                position = player.currentPosition
                duration = player.duration
            }

            val bufferedPosition = player.bufferedPosition.toInt()
            val isBuffering = player.playbackState == Player.STATE_BUFFERING

            if (duration > 0) {
                eventHandler.sendEvent("timeUpdate", mapOf(
                    "position" to position.toInt(),
                    "duration" to duration.toInt(),
                    "bufferedPosition" to bufferedPosition,
                    "isBuffering" to isBuffering
                ))
            }

            handler.postDelayed(this, UPDATE_INTERVAL_MS)
        }
    }

    // timeUpdateRunnable is started on the first STATE_READY transition
    // to avoid ticking every second before any media is loaded.

    fun release() {
        // Stop periodic updates
        if (isTimeUpdateRunning) {
            handler.removeCallbacks(timeUpdateRunnable)
            isTimeUpdateRunning = false
        }
        cancelStallWatchdog()
    }

    private fun startStallWatchdog() {
        cancelStallWatchdog()
        val runnable = Runnable {
            // playWhenReady=false means the user paused — recovery would
            // force-resume against their intent. onPlayWhenReadyChanged
            // re-arms the watchdog if they resume while still buffering.
            if (player.playbackState == Player.STATE_BUFFERING && player.playWhenReady) {
                stallRecoveryAttempt++

                if (stallRecoveryAttempt > MAX_RECOVERY_ATTEMPTS) {
                    Log.e(TAG, "Stall watchdog exceeded $MAX_RECOVERY_ATTEMPTS attempts — giving up")
                    eventHandler.sendEvent("error", mapOf("message" to "Playback stalled after $MAX_RECOVERY_ATTEMPTS recovery attempts"))
                    return@Runnable
                }

                hasReportedBuffering = false
                eventHandler.sendEvent("buffering")
                hasReportedBuffering = true

                if (stallRecoveryAttempt <= 1 || isInPipMode()) {
                    // Light recovery: re-prepare, preserving the media source.
                    // Always use light recovery during PiP — stop() is destructive and kills PiP.
                    // Only live streams get the default-position seek (the live
                    // edge); for VOD the default position is the window start
                    // and would yank the viewer back to 0:00.
                    Log.w(TAG, "Stall watchdog fired (attempt $stallRecoveryAttempt, pip=${isInPipMode()}) — light recovery")
                    if (player.isCurrentMediaItemLive) {
                        player.seekToDefaultPosition()
                    }
                    player.prepare()
                    player.play()
                } else {
                    // Heavy recovery: tear down and re-prepare from scratch.
                    Log.w(TAG, "Stall watchdog fired (attempt $stallRecoveryAttempt) — full stop/prepare/play")
                    player.stop()
                    player.prepare()
                    player.play()
                }

                // Re-arm for next attempt if still stuck
                startStallWatchdog()
            }
        }
        stallWatchdogRunnable = runnable
        handler.postDelayed(runnable, STALL_WATCHDOG_MS)
    }

    private fun cancelStallWatchdog() {
        stallWatchdogRunnable?.let { handler.removeCallbacks(it) }
        stallWatchdogRunnable = null
    }

    override fun onPlayWhenReadyChanged(playWhenReady: Boolean, reason: Int) {
        if (!playWhenReady) {
            // User paused — don't let the watchdog force-resume playback.
            cancelStallWatchdog()
        } else if (player.playbackState == Player.STATE_BUFFERING) {
            // Resumed while still buffering — restore stall detection.
            startStallWatchdog()
        }
    }

    override fun onPlaybackStateChanged(playbackState: Int) {
        Log.d(TAG, "Playback state changed: $playbackState, isLoading: ${player.isLoading}")
        when (playbackState) {
            Player.STATE_IDLE -> {
                cancelStallWatchdog()
            }
            Player.STATE_BUFFERING -> {
                // Start watchdog to detect indefinite buffering
                startStallWatchdog()
                // Send buffering event when entering BUFFERING state
                // Only send if we haven't already reported buffering
                if (!hasReportedBuffering) {
                    Log.d(TAG, "Entering BUFFERING state, sending buffering event")
                    eventHandler.sendEvent("buffering")
                    hasReportedBuffering = true
                }
            }
            Player.STATE_READY -> {
                cancelStallWatchdog()
                // Reset buffering flag and recovery counter when we're ready
                hasReportedBuffering = false
                stallRecoveryAttempt = 0

                // Start periodic time updates on first ready
                if (!isTimeUpdateRunning) {
                    handler.post(timeUpdateRunnable)
                    isTimeUpdateRunning = true
                }

                // Ready state is handled by onIsLoadingChanged when loading finishes
                // But send loaded event with duration here as it's state-specific
                val duration = player.duration.toInt()
                if (duration > 0 && !player.isLoading) {
                    eventHandler.sendEvent("loaded", mapOf("duration" to duration))
                }
            }
            Player.STATE_ENDED -> {
                cancelStallWatchdog()
                // When looping is enabled with REPEAT_MODE_ONE, this state shouldn't be reached
                // as ExoPlayer handles looping internally. However, handle it for safety.
                // Check actual repeat mode instead of stale enableLooping parameter
                if (player.repeatMode != Player.REPEAT_MODE_ONE) {
                    // Reset video to the beginning and pause
                    player.seekTo(0)
                    player.pause()
                    eventHandler.sendEvent("completed")
                }
                // Don't send completed event when looping (repeat mode is ON)
                // This ensures consistent behavior even if setLooping() was called after observer init
            }
        }
    }

    override fun onIsPlayingChanged(isPlaying: Boolean) {
        Log.d(TAG, "Is playing changed: $isPlaying, playbackState: ${player.playbackState}")
        if (isPlaying) {
            // ALWAYS update media session/notification when playback starts
            // This ensures media controls show the correct info whether in normal view or PiP
            val mediaInfo = getMediaInfo?.invoke()
            if (mediaInfo != null && notificationHandler != null) {
                val title = mediaInfo["title"] as? String
                Log.d(TAG, "📱 [Observer] Player started playing, updating media session for: $title")
                notificationHandler.setupMediaSession(mediaInfo)
                Log.d(TAG, "✅ [Observer] Media session updated - controls should now show correct info")
            } else {
                if (mediaInfo == null) {
                    Log.w(TAG, "⚠️ [Observer] No media info available when playing - media controls may not show correctly")
                }
                if (notificationHandler == null) {
                    Log.w(TAG, "⚠️ [Observer] No notification handler available")
                }
            }
            eventHandler.sendEvent("play")
        } else {
            // Only send pause event if not buffering
            // When seeking to unbuffered position, isPlaying becomes false but player is buffering
            // We should not report this as a pause - the buffering event will be sent instead
            if (player.playbackState != Player.STATE_BUFFERING) {
                eventHandler.sendEvent("pause")
            }
        }
    }

    override fun onIsLoadingChanged(isLoading: Boolean) {
        Log.d(TAG, "Is loading changed: $isLoading, playbackState: ${player.playbackState}, isPlaying: ${player.isPlaying}, playWhenReady: ${player.playWhenReady}")

        // Send buffering event when loading starts in BUFFERING state
        // This catches cases where isLoading changes before playbackState
        // Only send if we haven't already reported buffering
        if (isLoading && player.playbackState == Player.STATE_BUFFERING && !hasReportedBuffering) {
            Log.d(TAG, "Loading started while in BUFFERING state, sending buffering event")
            eventHandler.sendEvent("buffering")
            hasReportedBuffering = true
        } else if (!isLoading && player.playbackState == Player.STATE_READY) {
            // Reset buffering flag when loading finishes
            hasReportedBuffering = false

            // Note: "loaded" event is already sent by onPlaybackStateChanged when STATE_READY is reached
            // No need to send it again here

            // Restore the playback state after buffering completes
            // This tells the UI whether the video is playing or paused
            // IMPORTANT: Only send play/pause if player is not currently buffering
            // During initial buffering, isPlaying might be true (playWhenReady=true)
            // but the video hasn't actually started playing yet
            if (player.playbackState != Player.STATE_BUFFERING) {
                if (player.isPlaying) {
                    eventHandler.sendEvent("play")
                } else {
                    eventHandler.sendEvent("pause")
                }
            }
        }
    }

    override fun onPlayerError(error: PlaybackException) {
        if (error.errorCode == PlaybackException.ERROR_CODE_BEHIND_LIVE_WINDOW) {
            // Fell behind the live window — seek to live edge and re-prepare
            Log.w(TAG, "Behind live window, seeking to live edge")
            eventHandler.sendEvent("buffering")
            // Mark reported so the STATE_BUFFERING transition from prepare()
            // doesn't emit a duplicate buffering event.
            hasReportedBuffering = true
            player.seekToDefaultPosition()
            player.prepare()
            player.play()
        } else {
            Log.e(TAG, "Player error: ${error.message}", error)
            eventHandler.sendEvent(
                "error",
                mapOf("message" to (error.message ?: "Unknown error"))
            )
        }
    }

    override fun onDeviceInfoChanged(deviceInfo: androidx.media3.common.DeviceInfo) {
        // Check if playing to a remote device (Cast)
        val isExternalPlaybackActive = deviceInfo.playbackType == androidx.media3.common.DeviceInfo.PLAYBACK_TYPE_REMOTE

        // Only send event if the state changed
        if (isExternalPlaybackActive != wasExternalPlaybackActive) {
            wasExternalPlaybackActive = isExternalPlaybackActive
            Log.d(TAG, "Cast/external playback changed: $isExternalPlaybackActive")
            eventHandler.sendEvent(
                "airPlayConnectionChanged",
                mapOf("isConnected" to isExternalPlaybackActive)
            )
        }
    }
}
